File: src/example1/Solution5.java
package example1;

/**
 * @author Andrey Ledovskikh
 */
public class Solution5 {

    public static void main(String[] args) {
        var s = "bb";
        System.out.println(new Solution5().longestPalindrome(s));
    }

    public String longestPalindrome(String s) {
        if (s.length() == 1) {
            return s;
        }

        var maxPalindrome = s.substring(0, 1);
        for (var i = 0; i < s.length() - 1; i++) {
            if ((i + 1) < s.length() && s.charAt(i) == s.charAt(i + 1)) {
                var newP = cutPalindrome(s, i, i+1);
                maxPalindrome = newP.length() > maxPalindrome.length() ? newP : maxPalindrome;
            }

            if ((i + 2) < s.length() && s.charAt(i) == s.charAt(i + 2)) {
                var newP = cutPalindrome(s, i, i+2);
                maxPalindrome = newP.length() > maxPalindrome.length() ? newP : maxPalindrome;
            }
        }

        return maxPalindrome;
    }

    private String cutPalindrome(String s, int start, int end) {
        while (0 <= start && end < s.length() && s.charAt(start) == s.charAt(end) ) {
            start--;
            end++;
        }

        return s.substring(++start, end);
    }
}

File: js_src/Solution5.js
/**
 * @param {string} s
 * @return {string}
 */
const longestPalindrome = function(s) {
    if (s.length === 1) {
        return s;
    }

    let maxPalindrome = s[0];

    for (let i = 0; i < s.length - 1; i++) {
        if ((i + 1 < s.length) && s[i] === s[i + 1]) {
            let cur2 = curMaxPalindrome(s, i, i + 1);
            maxPalindrome = maxPalindrome.length < cur2.length ? cur2 : maxPalindrome;
        }

        if ((i + 2 < s.length) && s[i] === s[i + 2]) {
            let cur3 = curMaxPalindrome(s, i, i + 2);
            maxPalindrome = maxPalindrome.length < cur3.length ? cur3 : maxPalindrome;
        }
    }

    return maxPalindrome;
};

/**
 * @param {string} s
 * @param {int} start
 * @param {int} end
 * @returns {string}
 */
const curMaxPalindrome = (s, start, end) => {
    do {
        start--;
        end++;
    } while (0 <= start && end < s.length && s[start] === s[end])

    return s.substring(++start, end);
}

let s = "babad"
console.log(longestPalindrome(s));